### ISSUE TITLE
var, std memory consumption

### ISSUE THREAD

**Issue by 0x0L** [numpy/numpy#13199]
Hello,

It looks like numpy is allocating a whole new array to compute the second moment as seen in [_methods.py](https://github.com/numpy/numpy/blob/e6147b9bf580361f2f74ac72003f81e957587528/numpy/core/_methods.py#L117).

This could be easily improved by adding a low level function to compute the sum of squares (or any power for fast nth moment computation) along a dimension and expanding the square in `_var`.

It does not look too complicated but I'm really unsure on how to do the low level function (`loops.c.src` ?)

This should solve https://github.com/scikit-learn/scikit-learn/issues/5651

**Comment by ahaldane** [comment]
I think it's actually much easier than that: You just need to turn that line into an in-place operation. `np.asanyarray` doesn't make a copy, rather it is the subtraction operation that creates a new array.

So I think just replace `x = asanyarray(arr - arrmean)` by 
```python
arr -= arrmean
x = asanyarray(arr)
```
and that will avoid an allocation. Can you try that and see if it helps sklearn?

**Comment by eric-wieser** [comment]
`arr -= arrmean` will not work if `arr.dtype` is integral

**Comment by charris** [comment]
@ahaldane I suspect it was done that way for a reason. Note that `arr` and `arrmean` may be of different types, or so it looks to me on a quick scan.

**Comment by charris** [comment]
Note that it wasn't easy getting the function to work correctly for all the corner cases.

**Comment by 0x0L** [comment]
@ahaldane we don't want to modify the original array. Or am I missing something ?

**Comment by ahaldane** [comment]
@0x0L Ah, looking closer I see that you're right, we can't do an inplace there since it will modify the input array. The line you highlighted can't be inplace. In that case, which unnecessary memory allocation are you thinking of? For float64, everything appears to be inplace, including the sum of squares. 

For complex types, admittedly I think we're creating an array with the `np.conjugate` call.

**Comment by ahaldane** [comment]
Hmm, actually now maybe I see. What you mean is that we should have a loop (a ufunc) to do the entire operation `(arr - arrmean)**2` without any temporary array, right?

That's definitely doable as a ufunc. I want to think if it is possible without having to make a new ufunc.

**Comment by 0x0L** [comment]
We can use linalg.norm but it seems like an overkill

**Comment by ahaldane** [comment]
That seems really good. The only stubmbling block might be getting all the edge cases `np.var` currently accounts for (involving different dtypes and non-arrays) to work, but if we can get an implementation using that strategy to pass unit tests, it seems like a great improvement!

**Comment by dbarbier** [comment]
This approach is numerically unstable, see https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance.
There is also an interesting paper linked from https://github.com/dask/dask/pull/2301#issuecomment-309810317.  But a ufunc as described just above is surely simpler.

**Comment by ahaldane** [comment]
I did think about the ufunc. 

While it is technically possible (we can write any code, after all), numpy isn't really set up well for making one-off complicated ufuncs like that. Really it is designed around a set of single-operation core loops/ufuncs for multiplication, addition etc, and using buffers to carry out more complex combinations of these operations. This limitation is a big motivation for the numba project, which allows you to compile more complex inner loops easily. 

The limitation is also partly a motivation for ["generalized ufuncs"](https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html#c-api-for-implementing-elementary-functions) or "gufuncs" in numpy. If we really want to try to make this inner loop, that may be the way to go, in this case with signature `(i) -> ()`. Possible one should write it with signature `(i), () ->()` where the second argument would be the mean, calculated separately from the gufunc.

I have never written a gufunc myself, and I think that the gufunc code is relatively new in numpy and possibly still under development, so I'm not sure how hard it is. We may still be working on making gufuncs usable - I believe one of the major motivations was turning matmul into a gufunc, but that hasn't happened yet. So it may be that creating a (g)ufunc for this inner loop isn't practically feasible in numpy yet. 

Another dev more involved in gufuncs may be able to say better than me.

**Comment by mhvk** [comment]
If we go the `gufunc` route, we'd ideally optimize and use a 1-pass algorithm, e.g., as commented in a related issue (https://github.com/numpy/numpy/issues/9631#issuecomment-455712604), that of https://dl.acm.org/citation.cfm?id=3223036 (I quite enjoyed that paper).

The gufunc itself is not that hard to write, though it needs familiarizing oneself with the whole ufunc machinery; one could start from adapting something in https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/_umath_tests.c.src

Some possible issues: `axis=None` does not exist yet for `ufuncs`, so one would have to special-case that. Also when passing in, e.g., weights for a weighted mean and/or variance, you'll want to revisit the suggestion of allowing broadcasting over core dimensions...

**Comment by 0x0L** [comment]
Quick try at https://github.com/0x0L/numpy/commit/fd697e1e86d3e7708be62bc69c5eaf7b753daef9

```python
import numpy as np
from numpy.core._umath_tests import welford

a = np.random.randn(10, 5)

assert np.allclose(a.var(0), welford(a, axis=0))
assert np.allclose(a.var(1), welford(a, axis=1))
assert np.allclose(a.var(), welford(a.flat))
```

Not surprisingly, it's slower than the previous implementation for large vectors

**Comment by charris** [comment]
Note that Welford computes both the mean and variance in a single pass, so if you return both -- which is often wanted -- that might make up for the extra time in computing just the variance.

**Comment by mhvk** [comment]
@0x0L - that's cool!  It fixes #9631, and is substantially faster (factor 10!) for the small example array you gave.  For a large array, it is indeed a factor 3 slower, though:
```
a = np.random.normal(mean=1e7, size=(1000000,))
%timeit welford(a)
# 100 loops, best of 5: 6.94 ms per loop
%timeit a.var()
# 100 loops, best of 5: 2.24 ms per loop
```
Note that I think in part it is because the implementation is just a bit too complicated for vectorization: if I somewhat randomly remove the division, the addition to mu, or the addition of S, the time decreases to ~1 ms.

**Comment by mhvk** [comment]
p.s. For those without a subscription, the paper is also available directly from the author, at https://dbs.ifi.uni-heidelberg.de/files/Team/eschubert/publications/SSDBM18-covariance-authorcopy.pdf. Note that it briefly mentions that numpy makes an unnecessary copy...

**Comment by mhvk** [comment]
OK, got it down to being faster always, by using the Young and Cramer method referred to in the article (will PR to @0x0L branch shortly):
```
%timeit welford(a, axis=1)
# 1000000 loops, best of 5: 997 ns per loop
%timeit a.var(1)
# 10000 loops, best of 5: 15 µs per loop
a = np.random.normal(size=(1000000,))
%timeit welford(a)
# 100 loops, best of 5: 1.64 ms per loop
%timeit a.var()
# 100 loops, best of 5: 2.23 ms per loop
```

**Comment by 0x0L** [comment]
@mhvk Wow, that's great! For very large arrays the gain is ~ x4 for n = 5e8 on my system.

**Comment by seberg** [comment]
4x is pretty impressive, but I wonder how close we could get with streamlined two pass algorithm. I did not read the article, but it does say:

> Only when we can visit data only once (such as in streaming data, or when we
desire to use exponentially weighted moving averages) or when it
is expensive to visit twice (e.g., because it is too large to keep in
memory, or residing in a distributed storage) then incremental algo-
rithms such as ours are beneficial.

**Comment by mhvk** [comment]
@seberg - the algorithm generally wins for any case in which memory access dominates, so for large arrays. For other cases, the gains likely arise just from minimizing overheads.

Overall, my sense is that `.var()` is an operation that is done sufficiently frequently that optimization is worth it; in `C`, the `ndarray` method could even do the required ravel for `axis=None` before handing off to a `gufunc` (until `gufunc` get the ability to do this themselves...).

p.s. Do think that the `gufunc` should return both `mean` and `var`, since both are calculated.

**Comment by seberg** [comment]
@mhvk yes of course it is faster, but it does have some not necessarily super low precision loss as well, no? At least that was my understanding. I am completely good with moving it to C, although it might lose some maintenance simplicity I guess, that is probably worth it here. Returning both mean and var from the (or an internal gufunc) is good to.

**Comment by mhvk** [comment]
@seberg - for the precision, I guess there are both pluses and minuses - though sadly the faster version I made no longer fixes #9631:
```
x = [6715266981.538051]*10
np.array(x).var()
# 9.094947017729282e-13
welford(x)
# 5.820766091346741e-13
# For fun... (probably want double-precision integrand...)
welford(x, dtype='f4')
# 18641.352
np.array(x, dtype='f4').var()
# 262144.0
```

**Comment by seberg** [comment]
Well, I don't know, I had only a glance at the table, and while it was not bad compared to many others, that method was worse than the two pass. I just think possible precision loss is probably the only/largest hindrance to switching to a faster/different approach.

**Comment by charris** [comment]
If we want to go for high precision, the two pass Neely seems the best bet. Note that it also computes both mean and variance.  The AVX versions of cotton-cramers are very fast, I suspect that Neely could also be modified for AVX.

**Comment by mhvk** [comment]
Indeed, Table 2 in the paper has results for an AVX version for 2-pass Neely; overall, seems to be about 1.6 times slower than the single-pass for their 1e8 sample array. Neely has the best guarantees for accuracy even for bad data (small standard deviation around a high mean), but as you note, the case is pretty unlikely.

Of course, a gufunc would reduce memory and give a substantial speed-up no matter what algorithm one chooses.

**Comment by 0x0L** [comment]
I tried

```c
    BEGIN_OUTER_LOOP_2
        char *ip1=args[0], *op=args[1];
        @typ@ mean = 0;
        for (i = 0; i < di; i++, ip1 += is1) {
            mean += *(@typ@ *)ip1;
        }
        mean /= di;

        ip1 = args[0];
        @typ@ mu1=0, mu2=0;
        for (i = 0; i < di; i++, ip1 += is1) {
            const @typ@ x = *(@typ@ *)ip1 - mean;
            mu1 += x;
            mu2 += x * x;
        }
        mu1 /= di;
        mu2 = mu2 / di - mu1 * mu1;
        *(@typ@ *)op = mu2;
    END_OUTER_LOOP
```

It seems to solve #9631 (at least the example @mhvk gave)

On my machine for n = 5e8, I get
* current implem: 2.6s
* mhvk version 0.66s
* two-pass Neely 1s

**Comment by seberg** [comment]
Those are nice timings, aligns pretty well with the paper, I guess.

I find it hard to say what the 2 digits of precision mean in practice. Although, Neely solves some otherwise problematic cases as well I guess. I can be waived, but my preference would be to either stick with two-pass, or go for two-pass Neely.

It is a bit funny, if we implement Neely, the mean calculated/used by the variance function will – due to the second pass – sometimes be much more numerically stable than the normal mean function. An odd thought that might be possible (although I am not sure it is a good idea): With Neely one could probably detect and issue a warning in many cases that are Numerically problematic. Probably creates an explosion of complexity, but I wonder if you cannot use (partial) pairwise summation for `x * x` almost for free, and since `x * x` is always positive it could add a nice extra stability?

I simply prefer safe defaults to optimization. It would be awesome to have the speedy versions, but possibly that place should be outside of numpy proper for starters (a method kwarg may be something that could be discussed as well). If someone starts working on it and we are not quite sure where to go, we should maybe just invite the authors of that paper to comment on it.

**Comment by charris** [comment]
> I find it hard to say what the 2 digits of precision mean in practice.

Yes, one generally doesn't need high precision in the variance :) 

> Thus, except for the textbook approach,
any of the algorithms we evaluated is precise enough unless our
input data as in the example above has a very high dynamic range,
causing a loss of precision already when computing the sum.

Where I recall problems is the computation of the mean where the accumulation failed because of insufficient precision, roundoff error was simply discarding new data, the result wasn't a small loss of precision, but catastrophic failure. Neely might push the boundary for that a bit, but wouldn't solve it for all cases.

**Comment by mhvk** [comment]
In principle, after a single pass, you know whether the result is reasonably accurate, so one could even do a second pass as follow-up. But I'm quite OK with just going for a better two-pass; it still has the advantage of avoiding the intermediate array of differences.

**Comment by ahaldane** [comment]
Very interesting work all!

It's been alluded to above, but I wouldn't be opposed to putting a few sum/mean/var alternatives algorithms somewhere like in `numpy.lib`, with different tradeoffs in speed and accuracy. They seem like such useful fundamental operations that the cost/benefit of maintenance seems good. I personally have written my own module doing kahan variants of certain operations in the past, so I for one would have used them. But maybe I am getting ahead of things here, for now we are just deciding on how to implement `np.var`.

Also, the two-pass Neely sums seem like a good default balance in speed and accuracy to me as well, especially if it fixes other open issues.

**Comment by 0x0L** [comment]
Neely looks like a balanced default choice. Although it makes sense to have other variants in case the user care (in my ML work I would definitely go for speed) especially since it looks like it would be fairly simple to manage.

Some progress toward this at 0x0L@1cf3c77d469bfab3527afd7c695768b100b1ad79
Also started adding complex versions

**Comment by sk1p** [comment]
> I wouldn't be opposed to putting a few sum/mean/var alternatives algorithms somewhere like in numpy.lib, with different tradeoffs in speed and accuracy. 

:+1:  It would also be great if the "building blocks" for the streaming/incremental and parallel/distributed use cases, as mentioned in the paper above, were exposed.

**Comment by 0x0L** [comment]
While trying to reimplement `_var` and pass `runtests.py`, the biggest difficulty seems to be when `axis` is a tuple.

**Comment by mhvk** [comment]
A yes,  in `gufunc`s, the `axis` argument can only be a single integer - given that I implemented that, I should have remembered! The hope is to eventually have this be the same as for reductions, but we're not there yet... This means quite a bit of reshaping before and after (and having to take care with `out`...)

**Comment by 0x0L** [comment]
textbook one-pass and neely would be fine but Youngs-Cramer has `i` as state

**Comment by 0x0L** [comment]
I have a `_var` function more or less working at 0x0L@d0c6c6b5f459465cc03364964f33552aa3ef85b6

I went around the broadcasting issues by transposing a single axis. It's not pretty but it's still fairly simple and performs decently.

**Comment by RonaldAJ** [comment]
I have a slightly different proposal, but I lack the skills and insight in the source code to create the code for it:
https://mail.python.org/archives/list/numpy-discussion@python.org/message/Q222DZRWUL5RHZYAR4HCDSFJBA244FGC/

**Comment by endolith** [comment]
@ahaldane 

> but I wouldn't be opposed to putting a few sum/mean/var alternatives algorithms somewhere like in numpy.lib, with different tradeoffs in speed and accuracy

or a `method=` parameter to make the options more discoverable and document them all in one place?
